dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyPath.java

```diff
@@ -30,16 +30,29 @@
 public final class DependencyPath {
 
   private List<Artifact> path = new ArrayList<>();
+  private List<String> scopes = new ArrayList<>();
+  private List<Boolean> optionals = new ArrayList<>();
 
   @VisibleForTesting
-  public void add(Artifact artifact) {
+  public void add(Artifact artifact, String scope, Boolean optional) {
     path.add(artifact);
+    scopes.add(scope);
+    optionals.add(optional);
   }
   
   @Override
   public String toString() {
-    return Joiner.on(" / ")
-        .join(path.stream().map(Artifacts::toCoordinates).collect(Collectors.toList()));
+
+    ImmutableList.Builder<String> elements = ImmutableList.builder();
+    for (int i = 0; i < path.size(); ++i) {
+      Artifact artifact = path.get(i);
+      String scope = scopes.get(i);
+      boolean optional = optionals.get(i);
+      elements.add(
+          String.format(
```

Comment:
any reason not to use a local variable and string concatenation here?

Reply:
I like joiner to add something inbetween elements; I don't have to worry about removing a separator from tail or head.

Updated with DependencyPath.Node's toString.